opm/simulators/wells/WellHelpers.hpp

```diff
@@ -33,6 +40,108 @@ namespace Opm {
+        ///
+        /// For standard wells the B matrix, is basically a multiplication
+        /// of the equation of the perforated cells followed by a reduction
+        /// (summation) of these to the well equations.
+        ///
+        /// This class does that in the functions mv and mmv (from the DUNE
+        /// matrix interface.
+        ///
+        /// \tparam Scalar The scalar used for the computation.
+        template<typename Scalar>
+        class ParallelStandardWellB
+        {
+        public:
+            using Block = Dune::DynamicMatrix<Scalar>;
+            using Matrix = Dune::BCRSMatrix<Block>;
+
+            ParallelStandardWellB(const Matrix& B, const ParallelWellInfo& pinfo)
+                : B_(&B), pinfo_(&pinfo)
+            {}
+
+            //! y = A x
+            template<class X, class Y>
+            void mv (const X& x, Y& y) const
+            {
+    #if !defined(NDEBUG) && HAVE_MPI
+                // We need to make sure that all ranks are actually computing
+                // for the same well. Doing this by checking the name of the well.
+                int cstring_size = pinfo_->name().size()+1;
+                std::vector<int> sizes(pinfo_->communication().size());
+                pinfo_->communication().allgather(&cstring_size, 1, sizes.data());
+                std::vector<int> offsets(sizes.size()+1, 0); //last entry will be accumulated size
+                std::partial_sum(sizes.begin(), sizes.end(), offsets.begin() + 1);
+                std::vector<char> cstrings(offsets[sizes.size()]);
+                bool consistentWells = true;
+                char* send = const_cast<char*>(pinfo_->name().c_str());
+                pinfo_->communication().allgatherv(send, cstring_size,
+                                                   cstrings.data(), sizes.data(),
+                                                   offsets.data());
+                for(std::size_t i = 0; i < sizes.size(); ++i)
+                {
+                    std::string name(cstrings.data()+offsets[i]);
+                    if (name != pinfo_->name())
+                    {
+                        if (pinfo_->communication().rank() == 0)
+                        {
+                            //only one process per well logs, might not be 0 of MPI_COMM_WORLD, though
+                            std::string msg = std::string("Fatal Error: Not all ranks are computing for the same well")
+                                          + " well should be " + pinfo_->name() + " but is "
+                                + name;
+                            OpmLog::debug(msg);
+                        }
+                        consistentWells = false;
+                        break;
+                    }
+                }
+                pinfo_->communication().barrier();
+                // As not all processes are involved here we need to use MPI_Abort and hope MPI kills them all
```

Comment:
This should be dealt with in the BlackoilWellModel context (where all processes participate), using a DeferredLogger for the logging above, but as this is a pathological situation that (hopefully) is only arising during debugging, I think we can leave it as is for now.

Reply:
Sorry but that approach does not work anymore with distribributed wells if you want to have any chance of better performance.

Indeed the DeferredLogger will not work anymore as before as we are using different communicators. Maybe there is a fix, but certainly not at this point.

Reply:
> Sorry but that approach does not work anymore with distribributed wells if you want to have any chance of better performance.

I completely disagree. Are you saying that we cannot do proper communication of error conditions anymore? As long as we do not communicate such things too often, it should pose no problem. I think that for debugging it is ok, but in general it is not fine to crash and burn like this with MPI_Abort!

> Indeed the DeferredLogger will not work anymore as before as we are using different communicators. Maybe there is a fix, but certainly not at this point.

It will work just fine, it will still do communication with all processes. No change needed. As long as you push a message regarding a well to a DeferredLogger from just a single process (the well-set rank 0 as used above seems fine) all should be good. Then they could be collected at the global scope as before. We may want to avoid it for errors inside the apply() method, but that is an issue of granularity and avoiding communication too often, not of the DeferredLogger being unapplicable.

Reply:
The initializeWellPerfData() is also actually buggy, good example. I did not intend that you should remove this code, just tried to make sure we agree on the general approach to errors here.